src/crimson/osd/ops_executer.cc

```diff
@@ -723,6 +724,117 @@ version_t OpsExecuter::get_last_user_version() const
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+  // existing snaps are stored desendingly in snapc,
+  // cloned_snaps vector will hold all the snaps stored until snapset.seq
+   std::vector<snapid_t> cloned_snaps = [&] {
+      auto last = std::find_if(
+        std::begin(snapc.snaps), std::end(snapc.snaps),
+        [&](snapid_t snap_id) { return snap_id <= obc->ssc->snapset.seq; });
+      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
+    }();
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    auto snap_oi = prepare_clone(coid);
+
+    // make clone
+    do_write_op([this, snap_oi](auto& backend, auto& os, auto& txn) {
+      return backend.clone(snap_oi, os, clone_obc->obs, txn);
+    }, true);
+
+    delta_stats.num_objects++;
+    if (snap_oi->is_omap()) {
+      delta_stats.num_objects_omap++;
+    }
+    delta_stats.num_object_clones++;
+    // newsnapset is obc's ssc
+    obc->ssc->snapset.clones.push_back(coid.snap);
+    obc->ssc->snapset.clone_size[coid.snap] = obc->obs.oi.size;
+    obc->ssc->snapset.clone_snaps[coid.snap] = cloned_snaps;
+
+    // clone_overlap should contain an entry for each clone
+    // (an empty interval_set if there is no overlap)
+    auto &overlap = obc->ssc->snapset.clone_overlap[coid.snap];
+    if (obc->obs.oi.size) {
+      overlap.insert(0, obc->obs.oi.size);
+    }
+
+    // log clone
+    logger().debug("cloning v {} to {} v {} snaps= {} snapset={}",
+                   obc->obs.oi.version, coid,
+                   osd_op_params->at_version, cloned_snaps, obc->ssc->snapset);
+
+    log_entries.emplace_back(pg_log_entry_t::CLONE,
+                             coid, osd_op_params->at_version,
+                             obc->obs.oi.version, obc->obs.oi.user_version,
+                             osd_reqid_t(),
+			     obc->obs.oi.mtime, 0);
+    encode(cloned_snaps, log_entries.back().snaps);
+    osd_op_params->at_version.version++;
+
+    // TODO: update most recent clone_overlap and usage stats
+
+    if (snapc.seq > obc->ssc->snapset.seq) {
+       // update snapset with latest snap context
+       obc->ssc->snapset.seq = snapc.seq;
+       obc->ssc->snapset.snaps.clear();
+    }
+    logger().debug("{} {} done, snapset={}",
+      __func__, soid, obc->ssc->snapset);
+  }
+}
+
+object_info_t* OpsExecuter::prepare_clone(const hobject_t& coid)
+{
+    object_info_t static_snap_oi(coid);
+    object_info_t *snap_oi;
+    if (pg->is_primary()) {
+      // lookup_or_create
+      auto [c_obc, existed] =
+        pg->get_shard_services().obc_registry.get_cached_obc(
+          std::move(coid));
+      assert(!existed);
+      c_obc->obs.oi = static_snap_oi;
+      c_obc->obs.exists = true;
+      c_obc->ssc = obc->ssc;
+      c_obc->head = obc->head;
+      logger().debug("clone_obc: {}", c_obc->obs.oi);
+      clone_obc = std::move(c_obc);
+      snap_oi = &clone_obc->obs.oi;
+    } else {
+      snap_oi = &static_snap_oi;
```

Comment:
You can't return a pointer to an automatic variable, it will have been destructed before the caller sees it.

Reply:
You still seem to be returning a pointer to a stack variable.  Am I missing something?